qiskit/pulse/schedule.py

```diff
@@ -141,17 +141,39 @@ def _instructions(self, time: int = 0) -> Iterable[Tuple[int, 'Instruction']]:
         for insert_time, child_sched in self._children:
             yield from child_sched._instructions(time + insert_time)
 
-    def union(self, *schedules: List[ScheduleComponent],
+    def union(self, *schedules: Union[ScheduleComponent, Tuple[int, ScheduleComponent]],
               name: Optional[str] = None) -> 'Schedule':
-        """Return a new schedule which is the union of `self` and `schedule`.
+        """Return a new schedule which is the union of children of both `self` and `schedules`.
 
         Args:
             *schedules: Schedules to be take the union with this `Schedule`.
             name: Name of the new schedule. Defaults to name of self
         """
         if name is None:
             name = self.name
-        return Schedule(self, *schedules, name=name)
+        new_sched = Schedule(name=name)
+        new_sched._union((0, self))
+        for sched_pair in schedules:
+            if not isinstance(sched_pair, tuple):
+                sched_pair = (0, sched_pair)
+            new_sched._union(sched_pair)
+        return new_sched
+
+    def _union(self, other: Tuple[int, ScheduleComponent]) -> 'Schedule':
+        """Mutably union `self` and `other` Schedule with shift time.
+
+        Args:
+            other: Schedule with shift time to be take the union with this `Schedule`.
+        """
+        shift_time, sched = other
+        sched_timeslots = sched.timeslots
+        shifted_children = sched._children
+        if shift_time != 0:
```

Comment:
we can remove this check, since a schedule shifted by 0 is itself (what is the runtime of shift, now? maybe its too slow to not do this check?)

Reply:
Right, this is for performance. We can remove this check but it would make this function very slow, mainly due to the implementation of TimeslotCollection.shift (that takes O(n) now, it should be O(1) in the future). Also it's not bad to skip reconstruction of `shifted_children` for the case `other` has repeatedly many children, it may be rare, though.